cpp/include/cuml/experimental/fil/detail/evaluate_tree.hpp

```diff
@@ -0,0 +1,136 @@
+ * @tparam has_categorical nodes Whether or not this tree has any nodes with
+ * categorical splits
+ * @tparam node_t The type of nodes in this tree
+ * @tparam io_t The type used for input to and output from this tree (typically
+ * either floats or doubles)
+ * @param node Pointer to the root node of this tree
+ * @param row Pointer to the input data for this row
+ */
+template<
+  bool has_vector_leaves,
+  bool has_categorical_nodes,
+  typename node_t,
+  typename io_t
+>
+HOST DEVICE auto evaluate_tree(
+    node_t const* __restrict__ node,
+    io_t const* __restrict__ row
+) {
+  using categorical_set_type = bitset<uint32_t, typename node_t::index_type const>;
+  auto cur_node = *node;
+  do {
+    auto input_val = row[cur_node.feature_index()];
+    auto condition = true;
+    if constexpr (has_categorical_nodes) {
+      if (cur_node.is_categorical()) {
+        auto valid_categories = categorical_set_type{
+          &cur_node.index(),
+          uint32_t(sizeof(typename node_t::index_type) * 8)
+        };
+        condition = valid_categories.test(input_val);
+      } else {
+        condition = (input_val < cur_node.threshold());
+      }
+    } else {
+      condition = (input_val < cur_node.threshold());
+    }
+    if (!condition && cur_node.default_distant()) {
+      condition = isnan(input_val);
+    }
+    node += cur_node.child_offset(condition);
+    cur_node = *node;
+  } while (!cur_node.is_leaf());
+  return cur_node.template output<has_vector_leaves>();
+}
+
+/*
+ * Evaluate a single tree which requires external categorical storage on a
+ * single node
+ *
+ * For non-categorical models and models with a relatively small number of
+ * categories for any feature, all information necessary for model evaluation
+ * can be stored on a single node. If the number of categories for any
+ * feature exceeds the available space on a node, however, the
+ * categorical split data must be stored external to the node. We pass a
+ * pointer to this external data and reconstruct bitsets from it indicating
+ * the positive and negative categories for each categorical node.
+ *
+ * @tparam has_vector_leaves Whether or not this tree has vector leaves
+ * @tparam node_t The type of nodes in this tree
+ * @tparam io_t The type used for input to and output from this tree (typically
+ * either floats or doubles)
+ * @tparam categorical_storage_t The underlying type used for storing
+ * categorical data (typically char)
+ * @param node Pointer to the root node of this tree
+ * @param row Pointer to the input data for this row
+ * @param categorical_storage Pointer to where categorical split data is
+ * stored.
+ */
+template<
+  bool has_vector_leaves,
+  typename node_t,
+  typename io_t,
+  typename categorical_storage_t
+>
+HOST DEVICE auto evaluate_tree(
+    node_t const* __restrict__ node,
+    io_t const* __restrict__ row,
+    categorical_storage_t const* __restrict__ categorical_storage
```

Comment:
Can this kernel be called with a node that contains only local category storage?

After my first reading of the source code, here is my understanding:
* In `detail/decision_forest_builder.hpp`: Each categorical test node may use non-local storage or not. The decision is made once per node.
* In `detail/infer_kernel/cpu.hpp`: The kernel computes the flag `has_nonlocal_categories` only once for a given forest. The decision for using non-local storage is made once for the entire forest.

Is my understanding correct? If so, how do we ensure that we get consistent use of non-local storage for all categorical test node in a given forest? If some categorical nodes use non-local storage and others use local, then we may end up calling the wrong `evaluate_tree` kernel.

Reply:
I believe this is an actual bug. Will try to fix it tomorrow.

Reply:
Looking at it again, I think all is well, actually. The confusion (I believe) is with two variables with very similar names. At [line 93](https://github.com/wphicks/cuml/blob/fea-cpu_gpu_fil/cpp/include/cuml/experimental/fil/detail/decision_forest_builder.hpp#L93), we calculate the maximum number of categories for the current node (`max_node_categories`). Whether we proceed with non-local categories (line 94) is determined by the global maximum (`max_num_categories_`), not the node-specific maximum.

Reply:
Makes sense.